### ISSUE TITLE
CNN Attention based models in Keras

### ISSUE THREAD

**Issue by AntreasAntoniou** [fchollet/keras#1678]
Hello guys,

I've been trying to build an attention-based model using a CNN and an LSTM. Basically I want the LSTM to be able to select whichever slice it likes from the last layer of my CNN which is of size (512, 3, 3), so that each input will be of input 128, 3, 3 (using top 128 choices) transformed to 128, 9 for the LSTM. Now I thought of using an LSTM for the computations and one LSTM/Dense for selecting the slice and then sending the result as a softmax and the top results used as the slices. There is the possibility of using a Merge layer with mode='mul' but then I would have to have a matrix of the same dimensions which would not be very efficient. Is there any other ways around this in keras? Please let me know. 

Edit: I have looked into LamdaLayers, which may help. And also I thought of sending a node with 128 outputs as an input to my LSTM along with the conv outputs in hope that it will learn what each corresponds to. However I am open to any other hacks you guys may have come accross.

**Comment by Petrinet** [comment]
rom keras.layers import Merge
import numpy as np
np.random.seed(1337)  
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.layers import Embedding
from keras.layers import Convolution1D, GlobalMaxPooling1D
from keras.layers import LSTM
from keras.datasets import imdb
from keras import backend as K

max_features = 5000
maxlen = 400
batch_size = 32
embedding_dims = 50
nb_filter = 250
filter_length = 3
hidden_dims = 250
nb_epoch = 2

print('Loading data...')
(X_train, y_train), (X_test, y_test) = imdb.load_data(nb_words=max_features)
print(len(X_train), 'train sequences')
print(len(X_test), 'test sequences')

print('Pad sequences (samples x time)')
X_train = sequence.pad_sequences(X_train, maxlen=maxlen)
X_test = sequence.pad_sequences(X_test, maxlen=maxlen)
print('X_train shape:', X_train.shape)
print('X_test shape:', X_test.shape)

print('Build model...')
right_branch = Sequential()
right_branch.add(Embedding(max_features,
                    embedding_dims,
                    input_length=maxlen,
                    dropout=0.2))
right_branch.add(Convolution1D(
                        nb_filter=nb_filter,
                        filter_length=filter_length,
                        border_mode='valid',
                        activation='relu',
                        subsample_length=1))

left_branch = Sequential()
left_branch.add(Embedding(max_features,
                    embedding_dims,
                    input_length=maxlen,
                    dropout=0.2))
                 
left_branch.add(LSTM(398, dropout_W=0.5, dropout_U=0.5,return_sequences=False)) 


merged = Merge([left_branch,right_branch1], mode=dot,output_shape=lambda x: x[0])

final_model = Sequential()
final_model.add(merged)               
final_model.add(Dense(1))
final_model.add(Activation('sigmoid'))
final_model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
final_model.fit([X_train,X_train], y_train,
          batch_size=batch_size,
          nb_epoch=nb_epoch,
          validation_data=([X_test,X_test], y_test)
          )
rewrite merge dot

**Comment by agarwalanant** [comment]
Is this code working @Petrinet
